File: scripts/refresh-perm-data.py
import json
import re
import zipfile
from collections import Counter, defaultdict
from datetime import datetime, timedelta
from html import unescape
from pathlib import Path
from tempfile import NamedTemporaryFile
from urllib.request import Request, urlopen
from xml.etree.ElementTree import iterparse

PERFORMANCE_PAGE = "https://www.dol.gov/agencies/eta/foreign-labor/performance"
OUTPUT_PATH = Path(__file__).resolve().parents[1] / "public" / "perm-dashboard.json"
NS = "{http://schemas.openxmlformats.org/spreadsheetml/2006/main}"
HEADERS = {
    "User-Agent": "PERM-Dashboard-Importer/1.0 (public data dashboard)",
    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
    "Referer": "https://www.dol.gov/",
}


def fetch_bytes(url):
    with urlopen(Request(url, headers=HEADERS), timeout=180) as response:
        return response.read()


def newest_workbook_url(html):
    urls = [unescape(match.group(1)) for match in re.finditer(r'href=["\']([^"\']+\.xlsx(?:\?[^"\']*)?)["\']', html, re.I)]
    urls = [url if url.startswith("http") else "https://www.dol.gov" + url for url in urls]
    urls = [url for url in urls if "PERM" in url.upper() and "record_layout" not in url.lower()]
    return sorted(set(urls), key=lambda url: int(re.search(r"FY(20\d{2})", url, re.I).group(1)) if re.search(r"FY(20\d{2})", url, re.I) else 0)[-1]


def shared_strings(archive):
    values = []
    with archive.open("xl/sharedStrings.xml") as stream:
        for event, element in iterparse(stream, events=("end",)):
            if element.tag == f"{NS}si":
                values.append("".join(text.text or "" for text in element.iter(f"{NS}t")))
                element.clear()
    return values


def column_index(reference):
    letters = re.match(r"[A-Z]+", reference).group(0)
    result = 0
    for letter in letters:
        result = result * 26 + ord(letter) - 64
    return result - 1


def cell_value(cell, strings):
    value = cell.find(f"{NS}v")
    raw = "" if value is None else value.text or ""
    if cell.get("t") == "s":
        return strings[int(raw)] if raw else ""
    if cell.get("t") == "inlineStr":
        return "".join(text.text or "" for text in cell.iter(f"{NS}t"))
    return raw


def normalize(value):
    return re.sub(r"[^A-Z0-9]", "_", str(value or "").upper())


def parse_date(value):
    if not value:
        return None
    try:
        if re.fullmatch(r"\d+(\.\d+)?", str(value)):
            return datetime(1899, 12, 30) + timedelta(days=float(value))
        return datetime.fromisoformat(str(value).replace("Z", "+00:00")).replace(tzinfo=None)
    except ValueError:
        for pattern in ("%m/%d/%Y", "%m/%d/%y", "%Y-%m-%d"):
            try:
                return datetime.strptime(str(value), pattern)
            except ValueError:
                pass
    return None


def empty_state():
    return {"daily": Counter(), "weekly": Counter(), "monthly": Counter(), "weekday": Counter(), "letters": Counter(), "month_letters": defaultdict(Counter), "latest": None}


def add_record(state, date, status, employer):
    state["latest"] = date if state["latest"] is None or date > state["latest"] else state["latest"]
    state["daily"][date.date().isoformat()] += 1
    monday = date.date() - timedelta(days=date.weekday())
    state["weekly"][monday.isoformat()] += 1
    state["monthly"][date.strftime("%Y-%m")] += 1
    state["weekday"][date.weekday()] += 1
    letter = employer.strip()[:1].upper() if employer.strip()[:1].isalpha() else "X"
    state["letters"][letter] += 1
    state["month_letters"][date.strftime("%Y-%m")][letter] += 1


def build_series(state):
    latest = state["latest"] or datetime(1970, 1, 1)
    daily_dates = [latest - timedelta(days=29 - index) for index in range(30)]
    month_dates = [datetime(latest.year, latest.month, 1)]
    for _ in range(11):
        current = month_dates[0]
        month_dates.insert(0, datetime(current.year - (1 if current.month == 1 else 0), 12 if current.month == 1 else current.month - 1, 1))
    latest_month = latest.strftime("%Y-%m")
    weekly_starts = [latest.date() - timedelta(days=latest.weekday() + 21 - index * 7) for index in range(4)]
    return {
        "daily": [state["daily"][date.date().isoformat()] for date in daily_dates],
        "dailyLabels": [date.strftime("%b %-d") for date in daily_dates],
        "weekly": [state["weekday"][day] for day in range(7)],
        "weeklyLabels": ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"],
        "weeklyVolume": [state["weekly"][date.isoformat()] for date in weekly_starts],
        "weeklyVolumeLabels": [date.strftime("%b %-d") for date in weekly_starts],
        "monthlyVolume": [state["monthly"][date.strftime("%Y-%m")] for date in month_dates],
        "monthlyVolumeLabels": [date.strftime("%b %Y") for date in month_dates],
        "activity": [state["letters"][chr(65 + index)] for index in range(26)],
        "activeLetters": [state["month_letters"][latest_month][chr(65 + index)] for index in range(26)],
        "detailedAsOf": latest.isoformat() + "Z",
    }


def parse_workbook(path):
    processed = empty_state()
    certified = empty_state()
    with zipfile.ZipFile(path) as archive:
        strings = shared_strings(archive)
        with archive.open("xl/worksheets/sheet1.xml") as stream:
            header = None
            for event, row in iterparse(stream, events=("end",)):
                if row.tag != f"{NS}row":
                    continue
                values = {column_index(cell.get("r")): cell_value(cell, strings) for cell in row.findall(f"{NS}c")}
                if header is None:
                    candidate = {normalize(value): index for index, value in values.items()}
                    decision_column = next((candidate.get(name) for name in ("DECISION_DATE", "DATE_OF_DECISION", "CASE_DECISION_DATE", "DETERMINATION_DATE") if candidate.get(name) is not None), None)
                    if decision_column is not None:
                        header = (decision_column, candidate.get("CASE_STATUS", candidate.get("STATUS")), candidate.get("EMP_BUSINESS_NAME", candidate.get("EMPLOYER_NAME")))
                elif header:
                    date = parse_date(values.get(header[0]))
                    if date:
                        status = str(values.get(header[1], "")).upper()
                        employer = str(values.get(header[2], ""))
                        add_record(processed, date, status, employer)
                        if "CERTIFIED" in status or "CERTIFICATION" in status:
                            add_record(certified, date, status, employer)
                row.clear()
    if processed["latest"] is None:
        raise RuntimeError("The DOL workbook did not have decision rows with a usable DECISION_DATE column.")
    return processed, certified


html = fetch_bytes(PERFORMANCE_PAGE).decode("utf-8", errors="replace")
workbook_url = newest_workbook_url(html)
if not workbook_url:
    raise RuntimeError("No PERM disclosure workbook was found on the DOL performance page.")
with NamedTemporaryFile(suffix=".xlsx") as temporary:
    temporary.write(fetch_bytes(workbook_url))
    temporary.flush()
    processed, certified = parse_workbook(temporary.name)

OUTPUT_PATH.parent.mkdir(parents=True, exist_ok=True)
OUTPUT_PATH.write_text(json.dumps({
    "generatedAt": datetime.utcnow().isoformat() + "Z",
    "sourceUrl": workbook_url,
    "sourceRelease": workbook_url.rsplit("/", 1)[-1],
    "chartSeries": {"Certified": build_series(certified), "Processed": build_series(processed)},
}) + "\n")
print(json.dumps({"sourceUrl": workbook_url, "latestDecision": processed["latest"].isoformat()}))

File: scripts/refresh-perm-data.mjs

import { mkdir, writeFile } from "node:fs/promises";
import * as XLSX from "xlsx";

const PERFORMANCE_PAGE = "https://www.dol.gov/agencies/eta/foreign-labor/performance";
const OUTPUT_PATH = new URL("../public/perm-dashboard.json", import.meta.url);
const MAX_WORKBOOK_BYTES = 200 * 1024 * 1024;
const DOL_HEADERS = { "user-agent": "PERM-Dashboard-Importer/1.0 (public data dashboard)", "accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8", "referer": "https://www.dol.gov/" };

const landing = await fetch(PERFORMANCE_PAGE, { headers: DOL_HEADERS });
if (!landing.ok) throw new Error("DOL performance index failed (" + landing.status + ").");
const workbookUrl = newestDisclosureUrl(await landing.text());
if (!workbookUrl) throw new Error("No PERM disclosure workbook was found on the DOL performance page.");
const workbookResponse = await fetch(workbookUrl, { headers: DOL_HEADERS });
if (!workbookResponse.ok) throw new Error("DOL workbook download failed (" + workbookResponse.status + ").");
const size = Number(workbookResponse.headers.get("content-length") || 0);
if (size && size > MAX_WORKBOOK_BYTES) throw new Error("The DOL workbook exceeded the importer's configured size limit.");
const workbook = XLSX.read(Buffer.from(await workbookResponse.arrayBuffer()), { type: "buffer", cellDates: true });
const records = workbook.SheetNames.flatMap((sheetName) => readRecords(workbook.Sheets[sheetName]));
if (!records.length) throw new Error("The DOL workbook did not have decision rows with a usable DECISION_DATE column.");
const certified = records.filter((record) => /CERTIFIED|CERTIFICATION/.test(record.status));
const payload = { generatedAt: new Date().toISOString(), sourceUrl: workbookUrl, sourceRelease: workbookUrl.split("/").pop(), chartSeries: { Certified: buildChartSeries(certified), Processed: buildChartSeries(records) } };
await mkdir(new URL("../public/", import.meta.url), { recursive: true });
await writeFile(OUTPUT_PATH, JSON.stringify(payload) + String.fromCharCode(10));

function newestDisclosureUrl(html) {
  const urls = [...String(html).matchAll(/href=["']([^"']+[.]xlsx(?:[?][^"']*)?)["']/gi)].map((match) => new URL(match[1], PERFORMANCE_PAGE).href).filter((url) => /PERM/i.test(url) && !/record[ _-]?layout/i.test(url));
  return [...new Set(urls)].sort((left, right) => fiscalYear(right) - fiscalYear(left))[0];
}
function fiscalYear(url) { return Number(String(url).match(/FY(20[0-9]{2})/i)?.[1] || 0); }
function normalizeHeader(value) { return String(value || "").toUpperCase().replace(/[^A-Z0-9]/g, "_"); }
function firstColumn(columns, names) { return names.map((name) => columns[name]).find((value) => value !== undefined); }
function cellValue(sheet, rowIndex, columnIndex) { return columnIndex === undefined ? "" : sheet[XLSX.utils.encode_cell({ r: rowIndex, c: columnIndex })]?.v; }
function readRecords(sheet) {
  if (!sheet) return [];
  const range = XLSX.utils.decode_range(sheet["!ref"] || "A1:A1");
  const columns = {};
  for (let columnIndex = range.s.c; columnIndex <= range.e.c; columnIndex += 1) {
    const header = normalizeHeader(sheet[XLSX.utils.encode_cell({ r: range.s.r, c: columnIndex })]?.v);
    if (header) columns[header] = columnIndex;
  }
  const decisionColumn = firstColumn(columns, ["DECISION_DATE", "DATE_OF_DECISION", "CASE_DECISION_DATE", "DETERMINATION_DATE", "CASE_DETERMINATION_DATE", "FINAL_DECISION_DATE"]);
  const statusColumn = firstColumn(columns, ["CASE_STATUS", "STATUS", "FINAL_DECISION"]);
  const employerColumn = firstColumn(columns, ["EMPLOYER_NAME", "EMPLOYER_BUSINESS_NAME", "EMP_BUSINESS_NAME"]);
  if (decisionColumn === undefined) return [];
  const records = [];
  for (let rowIndex = range.s.r + 1; rowIndex <= range.e.r; rowIndex += 1) {
    const decisionDate = toDate(cellValue(sheet, rowIndex, decisionColumn));
    if (decisionDate) records.push({ decisionDate, status: String(cellValue(sheet, rowIndex, statusColumn) || "").toUpperCase(), employer: String(cellValue(sheet, rowIndex, employerColumn) || "") });
  }
  return records;
}
function toDate(value) {
  if (value instanceof Date && !Number.isNaN(value.getTime())) return value;
  if (typeof value === "number") { const parsed = XLSX.SSF.parse_date_code(value); return parsed ? new Date(parsed.y, parsed.m - 1, parsed.d) : null; }
  const date = new Date(value); return Number.isNaN(date.getTime()) ? null : date;
}
function buildChartSeries(records) {
  const latest = records.reduce((result, record) => record.decisionDate > result ? record.decisionDate : result, new Date(0));
  const dailyDates = Array.from({ length: 30 }, (_, index) => addDays(latest, index - 29));
  const monthDates = Array.from({ length: 12 }, (_, index) => new Date(latest.getFullYear(), latest.getMonth() - 11 + index, 1));
  const dailyCounts = countBy(records, (record) => dayKey(record.decisionDate));
  const weeklyCounts = countBy(records, (record) => weekKey(record.decisionDate));
  const monthlyCounts = countBy(records, (record) => monthKey(record.decisionDate));
  const letters = countBy(records, (record) => employerLetter(record.employer));
  const latestMonth = monthKey(latest);
  return { daily: dailyDates.map((date) => dailyCounts[dayKey(date)] || 0), dailyLabels: dailyDates.map(formatShortDate), weekly: [1, 2, 3, 4, 5, 6, 0].map((day) => records.filter((record) => record.decisionDate.getDay() === day).length), weeklyLabels: ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"], weeklyVolume: Array.from({ length: 4 }, (_, index) => weeklyCounts[weekKey(addDays(latest, -21 + index * 7))] || 0), weeklyVolumeLabels: Array.from({ length: 4 }, (_, index) => formatShortDate(addDays(latest, -21 + index * 7))), monthlyVolume: monthDates.map((date) => monthlyCounts[monthKey(date)] || 0), monthlyVolumeLabels: monthDates.map((date) => date.toLocaleDateString("en-US", { month: "short", year: "numeric" })), activity: "ABCDEFGHIJKLMNOPQRSTUVWXYZ".split("").map((letter) => letters[letter] || 0), activeLetters: "ABCDEFGHIJKLMNOPQRSTUVWXYZ".split("").map((letter) => records.filter((record) => monthKey(record.decisionDate) === latestMonth && employerLetter(record.employer) === letter).length), detailedAsOf: latest.toISOString() };
}
function countBy(items, keyForItem) { return items.reduce((counts, item) => { const key = keyForItem(item); counts[key] = (counts[key] || 0) + 1; return counts; }, {}); }
function addDays(date, amount) { const result = new Date(date); result.setDate(result.getDate() + amount); return result; }
function dayKey(date) { return date.toISOString().slice(0, 10); }
function weekKey(date) { return dayKey(addDays(date, -((date.getDay() + 6) % 7))); }
function monthKey(date) { return date.getFullYear() + "-" + String(date.getMonth() + 1).padStart(2, "0"); }
function formatShortDate(date) { return date.toLocaleDateString("en-US", { month: "short", day: "numeric" }); }
function employerLetter(value) { const letter = String(value || "").trim().charAt(0).toUpperCase(); return /[A-Z]/.test(letter) ? letter : "X"; }
